docs/design/2023-06-07-tiflash-pipeline-model.md

```diff
@@ -0,0 +1,72 @@
+# TiFlash pipeline model design doc
+
+* Author(s): [SeaRise](https://github.com/SeaRise), [ywqzzy](https://github.com/ywqzzy)
+* Tracking Issue: <https://github.com/pingcap/tiflash/issues/6518>
+
+## Table of Contents
+
+* [Motivation or Background](#motivation-or-background)
+* [Detailed Design](#detailed-design)
+* [Impacts & Risks](#impacts-risks)
+
+## Motivation or Background
+
+Currently, TiFlash's parallel execution model is a thread scheduling execution model, where each query will independently apply for several threads to perform collaborative execution.
+The thread scheduling model has two problems: 
+- In high-concurrency scenarios, too many threads will cause many context switches and incur high thread scheduling costs. When the number of threads reaches a certain number, the application of threads will report an error `thread constructor failed: Resource temporarily unavailable`. 
+- The thread scheduling model cannot accurately measure the resource usage of the query or perform fine-grained resource control.
+
+Although TiFlash has introduced several features to reduce the impact of high concurrency on the thread scheduling model, such as DynamicThreadPool, Async GRPC, and MinTsoScheduler, we can still improve the existing parallel execution model to better adapt to high-concurrency scenarios and support future resource control functions.
+
+Referring to [Morsel-Driven Parallelism: A NUMA-Aware Query Evaluation Framework for the Many-Core Age](https://15721.courses.cs.cmu.edu/spring2016/papers/p743-leis.pdf), we introduce a new concurrent execution model pipeline model and use a more efficient task scheduling mechanism.
+
+## Detailed Design
+
+![pipeline_model_overview](./images/2023-06-07-tiflash-pipeline-model.png)
+
+### Generate pipeline dag
+
+The plan tree sent to the query is divided into several pipelines according to the pipeline breaker and then assembled into a directed acyclic graph based on the dependency relationship. The definition of pipeline and pipeline breaker can refer to [Efficiently Compiling Efficient Query Plans for Modern Hardware](https://www.vldb.org/pvldb/vol4/p539-neumann.pdf).
+
+After the query is split into a pipeline dag, it will be submitted to the task scheduler for execution in sequence according to the DAG relationship.
```

Comment:
More details about how to handle the pipeline dependency in scheduling?

Reply:
ok, added some details
```
After the query is split into a pipeline dag, it will be submitted to the task scheduler for execution in sequence according to the DAG relationship.
The pipeline DAG is transformed into an event DAG where the dependencies between events correspond to the dependencies between pipelines.
Source events, i.e. events without upstream dependencies, are submitted directly to the task scheduler for execution.
After an event is completed, downstream events are triggered for execution.
An event is only submitted to the task scheduler for execution when all its upstream dependencies have completed.
             generate                submit              |
pipeline dag ────────►  event1 ────────────────────────► |
                                                         |
                          │  completed and then trigger  |
                          ▼                              |
                                     submit              |
                        event2 ────────────────────────► | task scheduler
                                                         |
                          │  completed and then trigger  |
                          ▼                              |
                                     submit              |
                        event3 ────────────────────────► |
```